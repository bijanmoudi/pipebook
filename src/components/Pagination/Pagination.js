import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import classnames from 'classnames';
import './index.scss';

const Pagination = ({
	currentPage,
	groupCount,
	hasNextPage,
	basePath,
	pathPageSelector,
	className,
	...restProps
}) => {
	const prevGroupPermitted = currentPage > groupCount;
	const prevPagePermitted = currentPage > 1;
	const nextPagePermitted = hasNextPage;
	const nextGroupPermitted = false; // Always disabled, because Pipedrive's API doesn't provide total count
	const currentPageIndex =
		currentPage % groupCount === 0 ? groupCount : currentPage % groupCount;
	const countOfPrevPagesInGroup = currentPageIndex - 1;
	const countOfNextPagesInGroup = groupCount - countOfPrevPagesInGroup - 1; // Again because of not having the total count, there will be a UI flash
	const getPath = page => {
		return basePath.replace(pathPageSelector, page);
	};
	return (
		<nav aria-label="Pagination">
			<ol
				className={classnames('pagination', {
					[className]: className
				})}
				{...restProps}
			>
				<li className="pagination__item pagination__item--previous-group">
					<Link
						to={
							prevGroupPermitted
								? getPath(currentPage - groupCount)
								: '#'
						}
						className={classnames('pagination__link', {
							'pagination__link--is-disabled': !prevGroupPermitted
						})}
						title="Previous set of pages"
						aria-disabled={!prevGroupPermitted}
						{...(prevGroupPermitted ? {} : { tabIndex: '-1' })}
					>
						<span className="visuallyhidden">
							Previous set of pages
						</span>
					</Link>
				</li>
				<li className="pagination__item pagination__item--previous-page">
					<Link
						to={prevPagePermitted ? getPath(currentPage - 1) : '#'}
						className={classnames('pagination__link', {
							'pagination__link--is-disabled': !prevPagePermitted
						})}
						title="Previous page"
						aria-disabled={!prevPagePermitted}
						{...(prevPagePermitted ? {} : { tabIndex: '-1' })}
					>
						<span className="visuallyhidden">Previous page</span>
					</Link>
				</li>
				{countOfPrevPagesInGroup
					? Array.apply(null, Array(countOfPrevPagesInGroup)).map(
							(item, i) => {
								const page =
									currentPage - countOfPrevPagesInGroup + i;
								return (
									<li
										key={`__prevpage${i}`}
										className="pagination__item"
									>
										<Link
											to={getPath(page)}
											className="pagination__link"
										>
											<span className="visuallyhidden">
												page #
											</span>
											{page}
										</Link>
									</li>
								);
							}
					  )
					: null}
				<li className="pagination__item pagination__item--is-active">
					<Link
						to={getPath(currentPage)}
						className="pagination__link"
						aria-current="page"
						tabIndex="-1"
					>
						<span className="visuallyhidden">page #</span>
						{currentPage}
					</Link>
				</li>
				{countOfNextPagesInGroup
					? Array.apply(null, Array(countOfNextPagesInGroup)).map(
							(item, i) => {
								const page = currentPage + i + 1;
								return (
									<li
										key={`__nextpage${i}`}
										className="pagination__item"
									>
										<Link
											to={
												nextPagePermitted
													? getPath(page)
													: '#'
											}
											className={classnames(
												'pagination__link',
												{
													'pagination__link--is-disabled': !nextPagePermitted
												}
											)}
											aria-disabled={!nextPagePermitted}
											{...(nextPagePermitted
												? {}
												: { tabIndex: '-1' })}
										>
											<span className="visuallyhidden">
												page #
											</span>
											{page}
										</Link>
									</li>
								);
							}
					  )
					: null}
				<li className="pagination__item pagination__item--next-page">
					<Link
						to={nextPagePermitted ? getPath(currentPage + 1) : '#'}
						className={classnames('pagination__link', {
							'pagination__link--is-disabled': !nextPagePermitted
						})}
						title="Next page"
						aria-disabled={!nextPagePermitted}
						{...(nextPagePermitted ? {} : { tabIndex: '-1' })}
					>
						<span className="visuallyhidden">Next page</span>
					</Link>
				</li>
				<li className="pagination__item pagination__item--next-group">
					<Link
						to={
							nextGroupPermitted
								? getPath(currentPage + groupCount)
								: '#'
						}
						className={classnames('pagination__link', {
							'pagination__link--is-disabled': !nextGroupPermitted
						})}
						title="Next set of pages"
						aria-disabled={!nextGroupPermitted}
						{...(nextGroupPermitted ? {} : { tabIndex: '-1' })}
					>
						<span className="visuallyhidden">
							Next set of pages
						</span>
					</Link>
				</li>
			</ol>
		</nav>
	);
};

Pagination.propTypes = {
	currentPage: PropTypes.number.isRequired,
	groupCount: PropTypes.number.isRequired,
	hasNextPage: PropTypes.bool.isRequired,
	basePath: PropTypes.string.isRequired,
	pathPageSelector: PropTypes.string.isRequired
};

Pagination.defaultProps = {
	currentPage: 1,
	groupCount: 3,
	hasNextPage: false,
	basePath: '/page/{page}',
	pathPageSelector: '{page}'
};

export default Pagination;
