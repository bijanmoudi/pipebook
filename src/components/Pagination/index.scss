/*=========================================
=            Pagination Styles            =
=========================================*/

@import '../../assets/styles/constants/layout';
@import '../../assets/styles/helpers/colors';
@import '../../assets/styles/helpers/units';

.pagination {
	display: flex;
	justify-content: center;
	align-items: center;
	user-select: none;
	&__item {
		& + * {
			margin-left: em(5, 14);
		}
		&--previous-group {
			.pagination__link {
				&:before {
					content: '«';
				}
			}
		}
		&--next-group {
			.pagination__link {
				&:before {
					content: '»';
				}
			}
		}
		&--previous-page {
			.pagination__link {
				&:before {
					content: '‹';
				}
			}
		}
		&--next-page {
			.pagination__link {
				&:before {
					content: '›';
				}
			}
		}
		&--is-active {
			.pagination__link {
				pointer-events: none;
				border-color: get-color('gray-dark');
				background-color: get-color('gray-dark');
				color: get-color('gray-light-ultra');
			}
		}
	}
	&__link {
		display: block;
		height: em(32, 14);
		min-width: em(32, 14);
		padding: 0 em(5, 14);
		border-radius: rem($border-radius);
		border: rem(2) solid get-color('gray-light');
		background-color: get-color('gray-light-ultra');
		color: get-color('gray-dark');
		text-decoration: none;
		text-align: center;
		line-height: em(28, 14);
		font-size: em(14);
		font-weight: $semi-bold-font-weight;
		&:before {
			font-size: em(18);
			line-height: 1;
		}
		&:hover,
		&:focus {
			border-color: get-color('gray-dark');
			color: get-color('gray-dark-very');
		}
		&--is-disabled {
			pointer-events: none;
			border-color: get-color('gray-light-very') !important;
			color: get-color('gray-light') !important;
		}
	}
}
