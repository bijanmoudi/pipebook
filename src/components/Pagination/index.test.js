import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import Pagination from './Pagination';

const props = {
	currentPage: 1,
	groupCount: 3,
	hasNextPage: false,
	basePath: '/page/{page}',
	pathPageSelector: '{page}'
};

describe('Testing "Pagination" component', () => {
	it('Rendering default pagination', () => {
		const tree = renderer
			.create(
				<Router>
					<Pagination {...props} />
				</Router>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
