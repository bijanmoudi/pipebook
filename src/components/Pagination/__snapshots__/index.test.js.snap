// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing "Pagination" component Rendering default pagination 1`] = `
<nav
  aria-label="Pagination"
>
  <ol
    className="pagination"
  >
    <li
      className="pagination__item pagination__item--previous-group"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
        title="Previous set of pages"
      >
        <span
          className="visuallyhidden"
        >
          Previous set of pages
        </span>
      </a>
    </li>
    <li
      className="pagination__item pagination__item--previous-page"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
        title="Previous page"
      >
        <span
          className="visuallyhidden"
        >
          Previous page
        </span>
      </a>
    </li>
    <li
      className="pagination__item pagination__item--is-active"
    >
      <a
        aria-current="page"
        className="pagination__link"
        href="/page/1"
        onClick={[Function]}
        tabIndex="-1"
      >
        <span
          className="visuallyhidden"
        >
          page #
        </span>
        1
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
      >
        <span
          className="visuallyhidden"
        >
          page #
        </span>
        2
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
      >
        <span
          className="visuallyhidden"
        >
          page #
        </span>
        3
      </a>
    </li>
    <li
      className="pagination__item pagination__item--next-page"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
        title="Next page"
      >
        <span
          className="visuallyhidden"
        >
          Next page
        </span>
      </a>
    </li>
    <li
      className="pagination__item pagination__item--next-group"
    >
      <a
        aria-disabled={true}
        className="pagination__link pagination__link--is-disabled"
        href="/"
        onClick={[Function]}
        tabIndex="-1"
        title="Next set of pages"
      >
        <span
          className="visuallyhidden"
        >
          Next set of pages
        </span>
      </a>
    </li>
  </ol>
</nav>
`;
